### PR TITLE
Feat/parse html from js template string

### PR DESCRIPTION
**Pull request by JeffersonCarvalh0**
I was facing a very similar problem to the one related in https://github.com/lukasgeiter/gettext-extractor/issues/21 so I decided to give it a try. The idea is to create a JS extractor that accepts an HTML parser as a parameter, in order to parse the HTML from template strings (pretty much the same idea from the `embeddedJs` extractor).
Please, let me know how I can improve this PR, and thanks for this amazing library!

PS: I'm not sure the linter worked

### CODE REVIEW

File: .npmignore
Comment (lukasgeiter) [inline .npmignore:1]: I don't think this file is needed

**Comment by lukasgeiter** [comment]
Thanks for your PR!

I'm wondering if limiting this feature to only template string literals with no substitutions is maybe a bit too restrictive? What are your thoughts on that?

**Comment by JeffersonCarvalh0** [comment]
I agree with you @lukasgeiter , I'm just not sure about which types I could check for (there are barely any docs for the compiler API from typescript - if you found a helpful source somewhere, please I'd like to know)

I could test with other types I find fitting in the `.d.ts` files from typescript and add them to this feature

**Comment by lukasgeiter** [comment]
No worries. I recommend you use a tool that makes it easy to explore the AST. Here's two I've used before:
- [astexplorer.net](https://astexplorer.net/)
- [ts-ast-viewer.com](https://ts-ast-viewer.com/)

For this feature there are two potential types of string literals to handle. First, regular string literals (double and single quotes). This one is quite straightforward: `SyntaxKind.StringLiteral`. Even better, use `isStringLiteralLike` to check for string literals and template strings without substitutions at once.

Secondly, there's the topic of template strings with substitutions. This one will be a bit more complex because you would first have to build a full string out of the parts that can then be passed to the html parser. Before you attempt that, I would be curious if you have this case in your code base. Just to get an idea if it's even worth it to implement support for template strings with substitutions.

**Comment by JeffersonCarvalh0** [comment]
Thanks for the tips!

As for html templates within template strings with substitutions, if I understood correctly, I actually have them in the code base, but they are already being extracted. My guess is that the JS extractor actually "sees" the JS code inside the template because it is already programmed to do that, regardless of the string surrounding it, so it doesn't matter if it is an HTML template (and the interpolated values don't generate more HTML code, they are literally calls to the i18n extraction function). I can't think about a use case where we'd have to write a specific logic to handle that

**Comment by JeffersonCarvalh0** [comment]
I've added `isStringLiteralLike` as suggested and also added a new test case to illustrate the cases I have in my codebase where template substitutions are used (note that I've added this test to the `callExpression` tests, not the `htmlTemplate` tests)

**Comment by lukasgeiter** [comment]
Yes, expressions inside template strings are JavaScript so they are properly extracted. There's no need for an explicit test for that. Otherwise I would have to add tests for every possible place a call expression can occur in JavaScript.

What I'm talking about is running the HTML parser/extractor on a template string with substitutions. For example:

```
`<div>
    <translate>Hello World</translate>
    ${substitution}
</div>`
```

File: tests/js/extractors/factories/htmlTemplate.test.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+      htmlParser = new HtmlParser(builder, [
+        (node, fileName, _, lineNumberStart) => {
+          if (node.nodeName === '#text') {
+            const textNode = node as DefaultTreeTextNode;
+            const startLineFromNode = textNode.sourceCodeLocation?.startLine;
+            const lineNumber = startLineFromNode ? lineNumberStart + startLineFromNode - 1 : lineNumberStart;
+            innerJsParser.parseString(textNode.value, fileName, { lineNumberStart: lineNumber });
+          }
+        }
```
Comment (lukasgeiter) [inline tests/js/extractors/factories/htmlTemplate.test.ts]: Could you please stick to one of the built in HTML extractors for the tests?

**Review by lukasgeiter — CHANGES_REQUESTED** [review]

**Comment by JeffersonCarvalh0** [comment]
> No worries. I recommend you use a tool that makes it easy to explore the AST. Here's two I've used before:
> 
> * [astexplorer.net](https://astexplorer.net/)
> * [ts-ast-viewer.com](https://ts-ast-viewer.com/)
> 
> For this feature there are two potential types of string literals to handle. First, regular string literals (double and single quotes). This one is quite straightforward: `SyntaxKind.StringLiteral`. Even better, use `isStringLiteralLike` to check for string literals and template strings without substitutions at once.
> 
> Secondly, there's the topic of template strings with substitutions. This one will be a bit more complex because you would first have to build a full string out of the parts that can then be passed to the html parser. Before you attempt that, I would be curious if you have this case in your code base. Just to get an idea if it's even worth it to implement support for template strings with substitutions.

Sorry about the delay.

Now I got what you meant. I don't think I have this case in my codebase =/
I've just applied the changes you suggested

File: tests/js/extractors/factories/callExpression.test.ts [hunk 2]
```diff
@@ -128,6 +128,31 @@ describe('JS: Call Expression Extractor', () => {
                 ]);
             });
 
+            test('HTML inside a template string with substitution', () => {
```
Comment (lukasgeiter) [inline tests/js/extractors/factories/callExpression.test.ts]: This test doesn't really add anything, please revert this change.

File: tests/js/extractors/factories/htmlTemplate.test.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    });
+
+    test('HTML inside a template literal', () => {
+      jsParser.parseString(`
+      #
+      #
+      #
+      #
+      #
+      #
```
Comment (lukasgeiter) [inline tests/js/extractors/factories/htmlTemplate.test.ts]: What's up with the `#`? That's not even valid JS

File: tests/js/extractors/factories/htmlTemplate.test.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      htmlParser = new HtmlParser(builder, [
+        HtmlExtractors.elementContent('translate')
+      ]);
+
+      jsParser = new JsParser(builder, [
+        htmlTemplateExtractor(htmlParser)
+      ]);
+    });
+
+    test('HTML inside a template literal', () => {
```
Comment (lukasgeiter) [inline tests/js/extractors/factories/htmlTemplate.test.ts]: I'd like to see some variety of tests. Take a look at `tests/html/extractors/factories/embeddedJs.test.ts`, you should be able to write very similar tests for this case.

Additionally, it would be good to have a test for each of the supported type of strings (regular string literal & template string literal).

Comment (lukasgeiter) [reply tests/js/extractors/factories/htmlTemplate.test.ts]: If you try to implement similar tests to the ones from embeddedJs, I expect `with lineNumberStart option` to be failing with this extractor. To resolve that, we basically have to implement this change for the js parser: #59

**Review by lukasgeiter — CHANGES_REQUESTED** [review]
Thanks for the changes!
Let's not bother with templates that have substitutions then. Support can still be added later if requested.

File: tests/js/extractors/factories/htmlTemplate.test.ts (hunk 3 above)
Comment (JeffersonCarvalh0) [reply tests/js/extractors/factories/htmlTemplate.test.ts]: Yeah, you're correct, I think I just had a brain fart while writing this one. My intention was to add line breaks to test if the parser would use the correct line numbers, but the `#` were not necessary indeed, simple line breaks would do the trick.

File: tests/js/extractors/factories/htmlTemplate.test.ts (hunk 4 above)
Comment (JeffersonCarvalh0) [reply tests/js/extractors/factories/htmlTemplate.test.ts]: I've added more tests, but I couldn't make the `separate line` one from `embeddedJs` work. I doesn't extract anything if I put `\n` inside the html

File: src/js/parser.ts [hunk 5]
```diff
@@ -3,7 +3,7 @@ import * as ts from 'typescript';
 import { Parser, IAddMessageCallback, IParseOptions } from '../parser';
 import { IMessage } from '../builder';
 
-export type IJsExtractorFunction = (node: ts.Node, sourceFile: ts.SourceFile, addMessage: IAddMessageCallback) => void;
+export type IJsExtractorFunction = (node: ts.Node, sourceFile: ts.SourceFile, addMessage: IAddMessageCallback, lineNumberStart?: number) => void;
```
Comment (lukasgeiter) [inline src/js/parser.ts]: Does `lineNumberStart` need to be optional? As far as I can tell, it should always be a number, no?

File: tests/js/extractors/factories/htmlTemplate.test.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      jsParser.parseString("let itBe = \"<div> <translate> test </translate> </div>\"");
+      expect(messages).toEqual([
+        {
+          text: 'test'
+        }
+      ])
+    });
+
+    test('single line (template string)', () => {
+      jsParser.parseString("let itBe = \"<div> <translate> test </translate> </div>\"");
```
Comment (lukasgeiter) [inline tests/js/extractors/factories/htmlTemplate.test.ts]: I'm liking the new tests! If you change the outer string to single quotes, you don't need to escape the double quotes: `'let itBe = "..."'`

File: tests/js/extractors/factories/htmlTemplate.test.ts (hunk 4 above)
Comment (lukasgeiter) [reply tests/js/extractors/factories/htmlTemplate.test.ts]: The separate line case is already covered by the other tests you have written, so that's fine.

**Review by lukasgeiter — CHANGES_REQUESTED** [review]
Thanks for the changes, and apologies for my delay. I think we're almost there 🙂

**Comment by JeffersonCarvalh0** [comment]
Sorry for the delay on my part, these last weeks have been busy =/
Please let me know if there is anything else that needs to be changed

**Comment by lukasgeiter** [comment]
Released with v3.8.0
